WIP - Schedule editor form widget

ryanmitchell opened the pull request:
Use of new JS widget and basic styling
Opening/closing hours translation to format, but no saving yet

Comments so far:
* Are we happy that this approach forces people into 15 minutes opening/closing slots?
* How do we handle multiple opening closing hours in the data structure, and where will this have knock on effects?

sampoyigi commented:
> Are we happy that this approach forces people into 15 minutes opening/closing slots?

I think that's fine as people will have an option between daily or flexible. If the slots are configurable we can make it an option?

> How do we handle multiple opening closing hours in the data structure, and where will this have knock on effects?

IIRC, WorkingSchedule class can already handle split working hours, we'll just need to adjust the workinghours unique index.

ryanmitchell commented:
I’ve split the slots into 15 minute increments in the JS - we could change it to different increments but obviously the number of boxes becomes quite large then.

> On 2 Oct 2020, at 19:47, Sam Poyigi <notifications@github.com> wrote:
> 
> 
> Are we happy that this approach forces people into 15 minutes opening/closing slots?
> 
> I think that's fine as people will have an option between daily or flexible. If the slots are configurable we can make it an option?
> 
> How do we handle multiple opening closing hours in the data structure, and where will this have knock on effects?
> 
> IIRC, WorkingSchedule class can already handle split working hours, we'll just need to adjust the workinghours unique index.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://github.com/tastyigniter/TastyIgniter/pull/500#issuecomment-702898999>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAAMVO2TKS3GROHKQNCZCNLSIYN4DANCNFSM4SAJR6WQ>.
>

sampoyigi commented:
15mins is fine then.

sampoyigi commented:
Few things to add
* A configurable option for the slots so it can be modified to an hour or 30 mins
* Can we also add the description column from the demo

sampoyigi commented:
Never mind on the configurable option.

ryanmitchell commented:
Ok. What do you want in the description? Does it need to be a text field?

> On 9 Oct 2020, at 18:43, Sam Poyigi <notifications@github.com> wrote:
> 
> ﻿
> Few things to add
> 
> A configurable option for the slots so it can be modified to an hour or 30 mins
> Can we also add the description column from the demo
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.

ryanmitchell commented:
Also not sure what you mean by multiple slots on a row with shift - you can click on the heading column to apply it to all. Plus click and mouse move selects in the direction of mouse movement until mouse up?

ryanmitchell commented:
Ignore the description question - I had never noticed on the demo it updates with the time selection, I'll bring that back.

sampoyigi commented:
> Also not sure what you mean by multiple slots on a row with shift - you can click on the heading column to apply it to all. Plus click and mouse move selects in the direction of mouse movement until mouse up?

Yea ignore this as well, the description column would do.

sampoyigi commented on app/admin/formwidgets/scheduleeditor/assets/js/scheduleeditor.js:
```diff
@@ -28,6 +28,116 @@
+                    data: {
+                        dimensions: [7,24*4],
+                        colHead: [
+                            {name:"00",title:"00:00-00:15"},{name:"&nbsp;&nbsp;&nbsp;&nbsp;",title:"00:15-00:30"},{name:"&nbsp;&nbsp;&nbsp;&nbsp;",title:"00:30-00:45"},{name:"&nbsp;&nbsp;&nbsp;&nbsp;",title:"00:45-01:00"},
+                            {name:"01",title:"01:00-01:15"},{name:"&nbsp;&nbsp;&nbsp;&nbsp;",title:"01:15-01:30"},{name:"&nbsp;&nbsp;&nbsp;&nbsp;",title:"01:30-01:45"},{name:"&nbsp;&nbsp;&nbsp;&nbsp;",title:"01:45-02:00"},
+                            {name:"02",title:"02:00-02:15"},{name:"&nbsp;&nbsp;&nbsp;&nbsp;",title:"02:15-02:30"},{name:"&nbsp;&nbsp;&nbsp;&nbsp;",title:"02:30-02:45"},{name:"&nbsp;&nbsp;&nbsp;&nbsp;",title:"02:45-03:00"},
+                            {name:"03",title:"03:00-03:15"},{name:"&nbsp;&nbsp;&nbsp;&nbsp;",title:"03:15-03:30"},{name:"&nbsp;&nbsp;&nbsp;&nbsp;",title:"03:30-03:45"},{name:"&nbsp;&nbsp;&nbsp;&nbsp;",title:"03:45-04:00"},
+                            {name:"04",title:"04:00-04:15"},{name:"&nbsp;&nbsp;&nbsp;&nbsp;",title:"04:15-04:30"},{name:"&nbsp;&nbsp;&nbsp;&nbsp;",title:"04:30-04:45"},{name:"&nbsp;&nbsp;&nbsp;&nbsp;",title:"04:45-05:00"},
+                            {name:"05",title:"05:00-05:15"},{name:"&nbsp;&nbsp;&nbsp;&nbsp;",title:"05:15-05:30"},{name:"&nbsp;&nbsp;&nbsp;&nbsp;",title:"05:30-05:45"},{name:"&nbsp;&nbsp;&nbsp;&nbsp;",title:"05:45-06:00"},
+                            {name:"06",title:"06:00-06:15"},{name:"&nbsp;&nbsp;&nbsp;&nbsp;",title:"06:15-06:30"},{name:"&nbsp;&nbsp;&nbsp;&nbsp;",title:"06:30-06:45"},{name:"&nbsp;&nbsp;&nbsp;&nbsp;",title:"06:45-07:00"},
+                            {name:"07",title:"07:00-07:15"},{name:"&nbsp;&nbsp;&nbsp;&nbsp;",title:"07:15-07:30"},{name:"&nbsp;&nbsp;&nbsp;&nbsp;",title:"07:30-07:45"},{name:"&nbsp;&nbsp;&nbsp;&nbsp;",title:"07:45-08:00"},
+                            {name:"08",title:"08:00-08:15"},{name:"&nbsp;&nbsp;&nbsp;&nbsp;",title:"08:15-08:30"},{name:"&nbsp;&nbsp;&nbsp;&nbsp;",title:"08:30-08:45"},{name:"&nbsp;&nbsp;&nbsp;&nbsp;",title:"08:45-09:00"},
+                            {name:"09",title:"09:00-09:15"},{name:"&nbsp;&nbsp;&nbsp;&nbsp;",title:"09:15-09:30"},{name:"&nbsp;&nbsp;&nbsp;&nbsp;",title:"09:30-09:45"},{name:"&nbsp;&nbsp;&nbsp;&nbsp;",title:"09:45-10:00"},
+                            {name:"10",title:"10:00-10:15"},{name:"&nbsp;&nbsp;&nbsp;&nbsp;",title:"10:15-10:30"},{name:"&nbsp;&nbsp;&nbsp;&nbsp;",title:"10:30-10:45"},{name:"&nbsp;&nbsp;&nbsp;&nbsp;",title:"10:45-11:00"},
+                            {name:"11",title:"11:00-11:15"},{name:"&nbsp;&nbsp;&nbsp;&nbsp;",title:"11:15-11:30"},{name:"&nbsp;&nbsp;&nbsp;&nbsp;",title:"11:30-11:45"},{name:"&nbsp;&nbsp;&nbsp;&nbsp;",title:"11:45-12:00"},
+                            {name:"12",title:"12:00-12:15"},{name:"&nbsp;&nbsp;&nbsp;&nbsp;",title:"12:15-12:30"},{name:"&nbsp;&nbsp;&nbsp;&nbsp;",title:"12:30-12:45"},{name:"&nbsp;&nbsp;&nbsp;&nbsp;",title:"12:45-13:00"},
+                            {name:"13",title:"13:00-13:15"},{name:"&nbsp;&nbsp;&nbsp;&nbsp;",title:"13:15-13:30"},{name:"&nbsp;&nbsp;&nbsp;&nbsp;",title:"13:30-13:45"},{name:"&nbsp;&nbsp;&nbsp;&nbsp;",title:"13:45-14:00"},
+                            {name:"14",title:"14:00-14:15"},{name:"&nbsp;&nbsp;&nbsp;&nbsp;",title:"14:15-14:30"},{name:"&nbsp;&nbsp;&nbsp;&nbsp;",title:"14:30-14:45"},{name:"&nbsp;&nbsp;&nbsp;&nbsp;",title:"14:45-15:00"},
+                            {name:"15",title:"15:00-15:15"},{name:"&nbsp;&nbsp;&nbsp;&nbsp;",title:"15:15-15:30"},{name:"&nbsp;&nbsp;&nbsp;&nbsp;",title:"15:30-15:45"},{name:"&nbsp;&nbsp;&nbsp;&nbsp;",title:"15:45-16:00"},
+                            {name:"16",title:"16:00-16:15"},{name:"&nbsp;&nbsp;&nbsp;&nbsp;",title:"16:15-16:30"},{name:"&nbsp;&nbsp;&nbsp;&nbsp;",title:"16:30-16:45"},{name:"&nbsp;&nbsp;&nbsp;&nbsp;",title:"16:45-17:00"},
+                            {name:"17",title:"17:00-17:15"},{name:"&nbsp;&nbsp;&nbsp;&nbsp;",title:"17:15-17:30"},{name:"&nbsp;&nbsp;&nbsp;&nbsp;",title:"17:30-17:45"},{name:"&nbsp;&nbsp;&nbsp;&nbsp;",title:"17:45-18:00"},
+                            {name:"18",title:"18:00-18:15"},{name:"&nbsp;&nbsp;&nbsp;&nbsp;",title:"18:15-18:30"},{name:"&nbsp;&nbsp;&nbsp;&nbsp;",title:"18:30-18:45"},{name:"&nbsp;&nbsp;&nbsp;&nbsp;",title:"18:45-19:00"},
+                            {name:"19",title:"19:00-19:15"},{name:"&nbsp;&nbsp;&nbsp;&nbsp;",title:"19:15-19:30"},{name:"&nbsp;&nbsp;&nbsp;&nbsp;",title:"19:30-19:45"},{name:"&nbsp;&nbsp;&nbsp;&nbsp;",title:"19:45-20:00"},
+                            {name:"20",title:"20:00-20:15"},{name:"&nbsp;&nbsp;&nbsp;&nbsp;",title:"20:15-20:30"},{name:"&nbsp;&nbsp;&nbsp;&nbsp;",title:"20:30-20:45"},{name:"&nbsp;&nbsp;&nbsp;&nbsp;",title:"20:45-21:00"},
+                            {name:"21",title:"21:00-21:15"},{name:"&nbsp;&nbsp;&nbsp;&nbsp;",title:"21:15-21:30"},{name:"&nbsp;&nbsp;&nbsp;&nbsp;",title:"21:30-21:45"},{name:"&nbsp;&nbsp;&nbsp;&nbsp;",title:"21:45-22:00"},
+                            {name:"22",title:"22:00-22:15"},{name:"&nbsp;&nbsp;&nbsp;&nbsp;",title:"22:15-22:30"},{name:"&nbsp;&nbsp;&nbsp;&nbsp;",title:"22:30-22:45"},{name:"&nbsp;&nbsp;&nbsp;&nbsp;",title:"22:45-23:00"},
+                            {name:"23",title:"23:00-23:15"},{name:"&nbsp;&nbsp;&nbsp;&nbsp;",title:"23:15-23:30"},{name:"&nbsp;&nbsp;&nbsp;&nbsp;",title:"23:30-23:45"},{name:"&nbsp;&nbsp;&nbsp;&nbsp;",title:"23:45-00:00"},
```
Can you generate these dynamically instead of hardcoding?

ryanmitchell replied:
Yeah I need to do a bit of refactoring now the functionality is in place

Will tidy it all up

sampoyigi reviewed (changes requested):
Also, use readable variable names.

ryanmitchell commented:
Ok, refactored it all to try and follow conventions from other JS plugins (events as separate functions, naming etc)

I've changed the data structure to be in the following format, to aid backwards compatibility:

[{
   "day": 0,
   "hours": [{
		open: "00:00",
		close: "01:00"
	}, {
		open: "02:30",
		close: "00:00"
	}],
   "status": 1,
},
...
]

sampoyigi reviewed (changes requested):
Great stuff with the refactor! I should be able to take over from here once these changes are resolved and come back if I need anything 👍

ryanmitchell commented on app/admin/formwidgets/scheduleeditor/assets/js/scheduleeditor.js:
```diff
@@ -154,6 +45,172 @@
+        
+        var data = [];
+        var initialRemarks = [];
+        for (var i=0; i<7; i++)
```
No problem - all replaced

sampoyigi replied:
Sweet, a lot more readable now 👍
